### PR TITLE
Enums with underlying types, because nice things are good (also parser boogfix)

### PR DESCRIPTION
**Pull request by alimpfard**
implementation is fairly complete, see README and sample.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -301,6 +301,20 @@ function look_at_foo(anonymous x: Foo) -> i32 {
...
+    UnexpectedMessage = 10
+    BadRecordMAC = 20
+    // etc
+}
+
+// Use in match:
+function do_nothing_in_particular() => match AlertDescription::CloseNotify {
+   CloseNotify => { ... }
+   UnexpectedMessage => { ... }
+   BadRecordMAC => { ... }
```
Comment (linusg) [inline README.md:316]: 4 spaces indentation please, even in the readme :D

Comment (alimpfard) [reply README.md:316]: oops~

File: src/typechecker.rs [hunk 2]
```diff
@@ -2805,20 +2899,21 @@ pub fn typecheck_expression(
...
+                                                                        Some(
+                                                                                substitute_typevars_in_type(
+                                                                                    field_type,
+                                                                                    &generic_parameters,
+                                                                                    project,
+                                                                                ),
+                                                                            )
+                                                                    } else {
+                                                                        field_type
+                                                                    };
```
Comment (linusg) [inline src/typechecker.rs:2916]: These appear to be identical, if that's really the case please consider moving this and the whitespace changes above into their own commit - not sure what's happening here, this both looks incredibly broken to me, but I'm sure rustfmt knows what it's doing :shrug:

Comment (alimpfard) [reply src/typechecker.rs:2916]: I'll see if I can make rustfmt not create this weirdo thing

**Comment by alimpfard** [comment]
Clippy errors are just the stuff from main :shrug:

**Comment by linusg** [comment]
I think one is yours actually, `main` has only one (which I just fixed)

**Comment by alimpfard** [comment]
Fixed that other one.

**Pull request merged**